malasakit-django/pcari/admin.py

```diff
@@ -146,7 +146,38 @@ def get_actions_wrapper(request):
 site.filter_actions(Group, ['delete_selected'])
 
 
-class ResponseAdmin(admin.ModelAdmin):
+class AdminViewMixin(admin.ModelAdmin):
+    """
+    Super class for admins to implement 'view' permissions.
+    """
+    def has_change_permission(self, request, obj=None):
+        """
+        This function is called to determine if a user can see objects in
+        the admin panel. Will return true if the user has either "change" or
+        "view" permissions on a model.
+        """
+        if admin.ModelAdmin.has_change_permission(self, request, obj):
+            return True
+        for perm in request.user.get_all_permissions():
+            if 'view_' + self.model.__name__.lower() == perm.split('.')[1]:
+                return True
+        return False
+
+    def get_readonly_fields(self, request, obj=None):
+        """
+        This function is called to determine which fields cannot be edited. If
+        a user only has "view" permissions, all fields are read-only. This does
+        not prevent data download.
+        """
+        if admin.ModelAdmin.has_change_permission(self, request, obj):
```

Comment:
Use `self.has_change_permission(request, obj)` (review how Python classes work). Repeat for all other places where this syntax occurs.

Reply:
I want to call the ```has_change_permission``` function of the super class because I want this to return true if the user has "change" permissions on an object. If I call ```self.has_change_permission``` It will return true if the user has either change or view permissions, but I want to return the default readonly fields when the user has change permissions to allow the user to change the default set of fields.